### PR TITLE
extended tagging.js to get all tagged to the current list

### PR DESCRIPTION
**Pull request by tobibeer**
Implements https://github.com/Jermolene/TiddlyWiki5/issues/1216#issuecomment-70498170
- `$:/core/modules/utils/utils.js`
  - introduces **$tw.utils.intersect** to retrieve an intersection of a variable number of arrays
- `$:/core/modules/filters/tagging.js`
  - updated filter implementing the suffix **tagging:all**

demo: http://tbdemo.tiddlyspot.com/#tagging%3Aall

### CODE REVIEW

**Comment by Jermolene** [comment]
Hi @tobibeer I'm not clear on the semantics of the intersection operation that you've defined. Can you expand the comment?

I'm not clear what the purpose is of `[tagging:all[]]`, but it seems like it should be a separate operator.

**Comment by tobibeer** [comment]
> I'm not clear on the semantics of the intersection operation that you've defined. Can you expand the comment?

[In set theory, an intersection is that which all sets have in common.](http://en.wikipedia.org/wiki/Intersection_%28set_theory%29) The purpose of **tagging:all** — which was also recently asked for in the groups by Danielo —  is pretty clear if you take a look at the bottom of the demo tiddler...

* **relevant / required for**
    * http://tagfilter.tiddlyspot.com

Here, you have a temporary tiddler that has been given certain tags and the list below is to show all tiddlers tagging precisely to the same tags as that temporary tiddler, i.e. similar to what tagsplorer / tagfiltr did for TWc.

Why would we need an additional filter when this is precisely a (variant of the) tagging filter?

**Comment by Jermolene** [comment]
Ok I'm inclined to think that "tagging:all" is quite a different operation than "tagging", and shouldn't be overloaded. One is a straightforward lookup of information, the other includes a comparison. Might it be possible to expose the underlying "intersect" operation as a filter? Perhaps one that returns the intersection between the current selection and a specified list-type field (ie, "list" or "tags")?

**Comment by tobibeer** [comment]
> One is a straightforward lookup of information, the other includes a comparison. 

Imho, both are straightforward lookups ...with respect to the same tagging relation...

* the current implementation prefers **or logic**
* the proposed implementation extends that with **and logic**

**$tw.utils.intersect** was only introduced as a global utility function so as to make it available for other usecases.

**Comment by Jermolene** [comment]
> Imho, both are straightforward lookups ...with respect to the same tagging relations...

Perhaps this is a case where writing some documentation that attempts to explain the feature would help: http://tiddlywiki.com/#FilterOperator%3A%20tagging

**Comment by tobibeer** [comment]
Yes. The importance to more thoroughly stamp out a spec / even documentation first is quite clear now.

**Comment by tobibeer** [comment]
not sure why / how those two additional files ended up here, tried to reset to HEAD of master but it won't do anything.

File: core/modules/utils/utils.js [hunk 1]
```diff
@@ -98,6 +98,31 @@ exports.pushTop = function(array,value) {
 };
 
 /*
+returns intersection of arrays
+*/
+exports.intersect = function() {
```
Comment (tobibeer) [inline core/modules/utils/utils.js]: @Jermolene, I have another filter now, that would need a function to return the [intersection](http://www.purplemath.com/modules/venndiag2.htm) of two attays.

**Comment by Jermolene** [comment]
Hi @tobibeer that's good! But I'd still want to postpone this until post-5.1.10. I've still got the concerns discussed above, plus I'm thinking that the intersection routine looks pretty inefficient (Googling "javascript intersect arrays" yields some better approaches).

**Comment by tobibeer** [comment]
> Googling "javascript intersect arrays" yields some better approaches.

Of course, there's always room for that. Btw., I was not hinting at this PR but only at that function ...afterwards actually thinking myself I should perhaps do some more research in terms of what's (perhaps) a more efficient routine... while pondering if and how to make use of objects with keys. Possibly the answer is: have that routine turn any arrays into objects before processing.

We could use a minified version of this:

http://pioupioum.fr/developpement/javascript-array-intersection.html

**Comment by tobibeer** [comment]
Closing in favor of the more recent #3003 by @inmysocks whose contains filter appears to cover the usecase introduced here.

**Pull request closed without merging**